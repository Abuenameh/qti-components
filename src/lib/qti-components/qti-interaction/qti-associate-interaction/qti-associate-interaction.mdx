import { Canvas, Meta } from '@storybook/blocks';

import * as Stories from './qti-associate-interaction.stories';

<Meta title="qti-interactions/qti-associate-interaction" />

# qti-associate-interaction

<Canvas of={Stories.Default} />
