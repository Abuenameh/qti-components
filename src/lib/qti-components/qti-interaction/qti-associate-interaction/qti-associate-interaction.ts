import { LitElement, css, html } from 'lit';

import { DragDropInteractionMixin } from '../internal/drag-drop';
import { state } from 'lit/decorators.js';

export class QtiAssociateInteraction extends DragDropInteractionMixin(
  LitElement,
  'qti-simple-associable-choice',
  true,
  '.dl'
) {
  @state() private _childrenMap: Element[];

  static override styles = css`
    :host {
      display: block; /* necessary to calculate scaling position */
    }
    slot[name='qti-simple-associable-choice'] {
      display: flex;
      flex-wrap: wrap;
      gap: 0.5rem;
    }
  `;

  override render() {
    return html` <slot name="prompt"></slot>
      <slot name="qti-simple-associable-choice"></slot>
      ${this._childrenMap.length > 0 &&
      Array.from(Array(Math.ceil(this._childrenMap.length / 2)).keys()).map(
        (_, index) =>
          html`<div part="associables-container">
            <div name="left${index}" part="drop-list" class="dl" identifier="droplist${index}_left"></div>
            <div name="right${index}" part="drop-list" class="dl" identifier="droplist${index}_right"></div>
          </div>`
      )}`;
  }

  override connectedCallback() {
    super.connectedCallback();
    this._childrenMap = Array.from(this.querySelectorAll('qti-simple-associable-choice'));
  }
}

customElements.define('qti-associate-interaction', QtiAssociateInteraction);
