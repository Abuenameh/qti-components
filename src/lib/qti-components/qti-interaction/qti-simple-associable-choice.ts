import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';

@customElement('qti-simple-associable-choice')
// tslint:disable: indent
export class QtiSimpleAssociableChoice extends LitElement {
  // pk: This needs some explanation
  // in the associate interaction there is a special slot for these qti-simple-associable-choices
  override connectedCallback(): void {
    super.connectedCallback();
    this.setAttribute('slot', 'qti-simple-associable-choice');
    this.setAttribute('part', 'qti-simple-associable-choice');
  }

  // pk: This needs some explanation
  // in qti-match-interaction, qti-simple-associable-choice is used to denote the
  // draggable, but also the droppable. WEIRD.. but lets deal with it.
  // So we have a slot for whenever another qti-simple-associable-choice is dropped in here.
  // And we have slot for content like in this associate interaction
  override render() {
    return html`
      <slot></slot>
      <slot name="qti-simple-associable-choice"></slot>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'qti-simple-associable-choice': QtiSimpleAssociableChoice;
  }
}
