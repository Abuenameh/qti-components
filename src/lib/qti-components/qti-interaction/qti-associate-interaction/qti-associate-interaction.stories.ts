import { action } from '@storybook/addon-actions';
import { html } from 'lit';

import './qti-associate-interaction';
import '../qti-simple-associable-choice';
import { ifDefined } from 'lit/directives/if-defined.js';

export default {
  component: 'qti-associate-interaction',
  argTypes: {
    'response-identifier': { control: { type: 'text' } },
    'min-associations': { control: { type: 'number' } },
    'max-associations': { control: { type: 'number' } },
    readonly: { control: { type: 'boolean' } },
    disabled: { control: { type: 'boolean' } },
    class: { control: 'inline-check', options: ['USA', 'Canada', 'asd', '12das', 'a2ersd', '34ads'] }
  }
};

export const Interaction = {
  render: args =>
    html`<qti-associate-interaction
      response-identifier=${ifDefined(args['response-identifier'])}
      @qti-register-interaction=${action(`qti-register-interaction`)}
      @qti-interaction-response=${action(`qti-interaction-response`)}
      class=${args.class}
      min-associations=${ifDefined(args['min-associations'])}
      max-associations=${ifDefined(args['max-associations'])}
      ?readonly=${args.readonly}
      ?disabled=${args.disabled}
    >
      <qti-prompt>
        Hidden in this list of characters from famous Shakespeare plays are three pairs of rivals. Can you match each
        character to his adversary?
      </qti-prompt>
      <qti-simple-associable-choice identifier="A" match-max="1">Antonio</qti-simple-associable-choice>
      <qti-simple-associable-choice identifier="C" match-max="1">Capulet</qti-simple-associable-choice>
      <qti-simple-associable-choice identifier="D" match-max="1">Demetrius</qti-simple-associable-choice>
      <qti-simple-associable-choice identifier="L" match-max="1">Lysander</qti-simple-associable-choice>
      <qti-simple-associable-choice identifier="M" match-max="1">Montague</qti-simple-associable-choice>
      <qti-simple-associable-choice identifier="P" match-max="1">Prospero</qti-simple-associable-choice>
    </qti-associate-interaction>`
};
