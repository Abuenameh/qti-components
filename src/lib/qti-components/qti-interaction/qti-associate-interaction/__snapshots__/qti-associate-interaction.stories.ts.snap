// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/qti-components/qti-interaction/qti-associate-interaction Default smoke-test 1`] = `
<div id="root-inner">
  <qti-associate-interaction class
                             min-associations="1"
                             max-associations="1"
  >
    <qti-prompt>
      Hidden in this list of characters from famous Shakespeare plays are three pairs of rivals. Can you match each
        character to his adversary?
    </qti-prompt>
    <qti-simple-associable-choice identifier="A"
                                  match-max="1"
                                  qti-draggable="true"
                                  slot="qti-simple-associable-choice"
                                  part="qti-simple-associable-choice"
    >
      Antonio
    </qti-simple-associable-choice>
    <qti-simple-associable-choice identifier="C"
                                  match-max="1"
                                  qti-draggable="true"
                                  slot="qti-simple-associable-choice"
                                  part="qti-simple-associable-choice"
    >
      Capulet
    </qti-simple-associable-choice>
    <qti-simple-associable-choice identifier="D"
                                  match-max="1"
                                  qti-draggable="true"
                                  slot="qti-simple-associable-choice"
                                  part="qti-simple-associable-choice"
    >
      Demetrius
    </qti-simple-associable-choice>
    <qti-simple-associable-choice identifier="L"
                                  match-max="1"
                                  qti-draggable="true"
                                  slot="qti-simple-associable-choice"
                                  part="qti-simple-associable-choice"
    >
      Lysander
    </qti-simple-associable-choice>
    <qti-simple-associable-choice identifier="M"
                                  match-max="1"
                                  qti-draggable="true"
                                  slot="qti-simple-associable-choice"
                                  part="qti-simple-associable-choice"
    >
      Montague
    </qti-simple-associable-choice>
    <qti-simple-associable-choice identifier="P"
                                  match-max="1"
                                  qti-draggable="true"
                                  slot="qti-simple-associable-choice"
                                  part="qti-simple-associable-choice"
    >
      Prospero
    </qti-simple-associable-choice>
  </qti-associate-interaction>
</div>
`;
